python/grass/pygrass/vector/abstract.py

```diff
@@ -413,14 +413,33 @@ def open(
             str_err = "Not able to open the map, C function return %d."
             raise OpenError(str_err % openvect)
 
+        # Load attribute table for selected layer.
         if len(self.dblinks) == 0:
             self.layer = layer
             self.table = None
             self.n_lines = 0
         else:
-            self.layer = self.dblinks.by_layer(layer).layer
-            self.table = self.dblinks.by_layer(layer).table()
+            layer_db_link = self.dblinks.by_layer(layer)
+            if not layer_db_link:
+                raise LookupError(
+                    "There appears to be no database link for layer %d of <%s>."
+                    % (layer, self.name)
+                )
+            if layer_db_link.layer != layer:
+                raise RuntimeError(
+                    "The databse link for layer %d of <%s> references layer %d."
+                    % (layer, self.name, layer_db_link.layer)
+                )
```

Comment:
This is a new exception; the previous code would just set `self.layer = layer_db_link.layer`. I don't know if there is a scenario where one would want to allow such a layer number mismatch.